theia-extensions/viewer-prototype/src/node/trace-server-service.ts

```diff
@@ -1,60 +1,88 @@
+            return new Promise<never>((_, reject) => server.once('error', reject));
+        }
+        this.server = server;
+        this.port = port;
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        let timeout: any;
+        try {
+            await new Promise<void>((resolve, reject) => {
+                // If the server doesn't error or exit within 2 seconds, we consider that a success.
+                // That doesn't mean that the server has really started. On the frontend, the `TraceServerConnectionStatusService`
+                // will ping the port until it receives a response, which is the official measure of success.
+                timeout = setTimeout(resolve, 2000);
+                // If the server exits or errors before it outputs, consider it a failure.
+                server.once('exit', (code, _signal) => reject(PortBusy(code)));
+                server.once('error', reject);
             });
-            // If no response recieved from the trace server in 10 seconds, reject with an error (for internal use)
-            timeouts.push(setTimeout(() => reject('Waited 10 seconds but nothing happened and hence exiting'), 10000));
-
-        }).finally(() => {
+        } finally {
             server.removeAllListeners();
-            timeouts.forEach(timeout => clearTimeout(timeout));
-        });
-
+            // The process exited so we can clean our internal state.
+            server.once('exit', () => {
+                this.server = undefined;
+                this.port = undefined;
+            });
+            clearTimeout(timeout);
+        }
+        return SUCCESS;
     }
 
-    stopTraceServer(): Promise<string> {
-        return new Promise<string>((resolve, reject) => {
-            if (this.processId === -1 || !this.processId) {
-                reject('process already killed or process hasn\'t been started');
-                return;
-            }
-            treeKill(this.processId, error => {
-                this.processId = -1;
+    async stopTraceServer(): Promise<string> {
+        if (!this.isServerRunning(this.server)) {
+            return SUCCESS;
+        }
+        const exit = new Promise<void>(resolve => {
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            this.server!.once('exit', resolve);
+        });
+        await new Promise<void>((resolve, reject) => {
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            treeKill(this.server!.pid!, error => {
                 if (error) {
                     reject(error);
-                    return;
+                } else {
+                    resolve();
                 }
-                resolve('success');
             });
         });
+        const timeout = new Promise((_, reject) => {
+            setTimeout(reject, 1000, new Error('the Trace Server did not exit'));
+        });
```

Comment:
The Promise logic here is a little odd:
 - Create a Promise that checks for exit
 - `await` a Promise that actually does the exiting. If this succeeds, aren't we guaranteed to have resolved the first Promise?
 - Create a new Promise with a timeout - we only get here if we successfully `await`ed `treeKill`, so we've probably already exited?
 - Race promises 1 & 3

Some comments might be useful to understand why it's done this way, if this is the right way to do it.

Reply:
It is to use `server.on('exit', ...)` as source of truth: As you said `treeKill` should kill the process, so `exit` should be fired. Since we relay on that `exit` event to clear the state of the class, we want to make sure that it is correctly invoked.